code/game/objects/items/devices/gps.dm

```diff
@@ -1,28 +1,423 @@
+		I.color = screen_color
+		I.appearance_flags |= RESET_COLOR
+		add_overlay(I)
+
+/obj/item/gps/attack_self(mob/user)
+	display(user)
+
+ // Compiles all the data not available directly from the GPS
+ // Like the positions and directions to all other GPS units
+/obj/item/gps/proc/display_list()
+	var/list/dat = list()
+
+	var/turf/curr = get_turf(src)
+	var/area/my_area = get_area(src)
+
+	dat["my_area_name"] = strip_improper(my_area.name)
+	dat["curr_x"] = curr.x
+	dat["curr_y"] = curr.y
+	dat["curr_z"] = curr.z
+	dat["curr_z_name"] = strip_improper(using_map.get_zlevel_name(curr.z))
+
+	if(long_range)
+		dat["z_level_detection"] = (global.using_map.station_levels|global.using_map.contact_levels|global.using_map.player_levels)
+	else
+		dat["z_level_detection"] = GetConnectedZlevels(curr.z)
+	
+	var/list/gps_list = list()
+	for(var/obj/item/gps/G as anything in global.all_gps_units)
+
+		if(G == src || !can_track(G, dat["z_level_detection"]))
+			continue
+
+		var/gps_data[0]
+		gps_data["ref"] = weakref(G)
+		gps_data["gps_tag"] = G.gps_tag
+
+		var/area/A = get_area(G)
+		gps_data["area_name"] = strip_improper(A.name)
+
+		var/turf/T = get_turf(G)
+		gps_data["z_name"] = strip_improper(using_map.get_zlevel_name(T.z))
+		gps_data["direction"] = get_compass_direction_string(curr, T)
+		gps_data["degrees"] = round(Get_Angle(curr,T))
+		gps_data["distX"] = T.x - curr.x
+		gps_data["distY"] = T.y - curr.y
+		gps_data["distance"] = get_dist(curr, T)
+		gps_data["local"] = (curr.z == T.z)
+		gps_data["x"] = T.x
+		gps_data["y"] = T.y
+
+		gps_list[++gps_list.len] = gps_data
+
+	dat["gps_list"] = gps_list
+
+	return dat
+
+/obj/item/gps/proc/display(mob/user)
+
+	if(emped)
+		to_chat(user, SPAN_WARNING("It's busted!"))
+		return
+
+	var/list/dat = list()
+	var/list/gps_data = display_list()
+
+	dat += "<table width = '100%'>"
+	if(!tracking)
+		dat += "<tr><td colspan = 6></td><a href='?src=\ref[src];toggle_power=1'>Switch On</a></tr>"
+	else
+		dat += "<tr><td colspan = 6></td><a href='?src=\ref[src];toggle_power=1'>Switch Off</a></tr>"
+		dat += "<tr><td colspan = 2><b>Current location</b></td><td colspan = 2>[gps_data["my_area_name"]]</td><td colspan = 2><b>([gps_data["curr_x"]], [gps_data["curr_y"]], [gps_data["curr_z_name"]])</b></td></tr>"
+		dat += "<tr><td colspan = 4>[hide_signal ? "Tagged" : "Broadcasting"] as '[gps_tag]'.</td>"
+		dat += "<td><a href='?src=\ref[src];tag=1'>Change Tag</a><a href='?src=\ref[src];range=1'>Toggle Scan Range</a>[can_hide_signal ? "<a href='?src=\ref[src];hide=1'>Toggle Signal Visibility</a>":""]</td></tr>"
+
+		var/list/gps_list = gps_data["gps_list"]
+		if(gps_list.len)
+			dat += "<tr><td colspan = 6><b>Detected signals</b></td></tr>"
+			for(var/gps in gps_data["gps_list"])
+
+				var/weakref/gps_weakref = gps["ref"]
+				if(!istype(gps_weakref))
+					continue
+				var/obj/item/gps/gps_obj = gps_weakref.resolve()
+				if(!istype(gps_obj) || QDELETED(gps_obj))
+					continue
```

Comment:
I don't get it. We literally just created this weakref in the helper at the head of this proc, and now we unwrap it synchronously here? Why?

Reply:
To be honest all of the display stuff is something I tidied a little on Polaris and then ported over. I may just write a NanoUI template and burn it to the ground.

Reply:
I think the UI stuff is weird but wouldn't block over most of it--with this weakref thing being the exception.

Reply:
Gutted it.